Migrate quarkus-oidc to Vert.x Mutiny WebClient

sberyozkin opened the pull request:
Fixes #14492.
Fixes #15004.
Fixes #12984.
Fixes #4553.

Also Fixes #15336.

(few more opened issues will be resolved in a follow up PR - otherwise this PR will become very massive).

This PR does the following:

1. Migrates `quarkus-oidc` to `Vert.x Mutiny Web Client`. The required changes have not been hard to implement with:
 - Mutiny WebClient is already returning `Uni`.
 - OIDC provider is represented as `OidcProvider` and the client as `OidcProviderClient`. `OidcProvider` does the locat verification, JWK refresh, the client - all the remote calls.
 - Client JWT authentication is now consistently supported at all levels, without specific workarounds
 - Discovery info is captured as `OidcConfigurationMetadata` - it will also help with resolving the issue @danielpetisme has asked about (injecting it in the endpoint)
- JWK refresh is now not decoupled from the current request which fixes #15004. 
- Introspection is still done by default if the JWK verification fails - but it can be disabled which also fixes #12984 where one can cause some excessive remote introspection calls.
- Specific reasons for the verification failures are now logged such that they can be easily captured (#4553)
- etc
2. Tests - none of the existing tests have been tweaked to get them passing, but also some of them improved as follows:
  - `integration-tests/oidc-tenancy` - there you can see that now there is no await is needed to get an expected successful verification of the token when JWK refresh is required (this is #15004) and also it is now possible to check a specific number of times the introspection endpoint has been called. A new test has been added to verify the remote introspection can be disabled.
 - Now Jose4j validates the token claims at the moment it is asked to verify it - so I removed now redundant code where token JSON is checled for the issuer/audience and the not very useful tests (only that utility code is checked) but instead updated `oidc-wiremock` tests
3. Minor updates to oidc client - to get a smaller exception trace with an easier to see message when the connection times out

Once this rather large PR gets merged I'll be able to look at rebasing the Vertx.4 branch against it, so it would be good if it could be reviewed soon. It can't be perfect from the 1st try :-), but we will get there in a few rounds :-)

Generally speaking I'm quite happy we will now be able to control much easier the way the OIDC logic flows.
As I said, a few more issue fixes will follow asap after this PR, and one particular optimizaition would be about avoiding some duplicate token parsing which is now realistic to do. Longer term - there is an obvious opportunity to have a common OIDC client code, etc

sberyozkin committed:
Migrate quarkus-oidc to Vert.x Mutiny WebClient

gsmet commented:
Just a note that you might need these 2 dependencies somewhere: https://github.com/quarkusio/quarkus/pull/15107/files .

sberyozkin commented:
@gsmet Hi, oh thanks, I'll add then to `oidc-common`.

@cescoffier @pedroigor @stuartwdouglas This migration is now complete in its first iteration. I'll be cleaning up some code next week - PR should be ready for review soon...
thanks

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:
```diff
@@ -232,80 +241,100 @@ private static JsonObject getRolesJson(RoutingContext vertxContext, TenantConfig
-            return Uni.createFrom().failure(new AuthenticationFailedException(ex));
-        }
+        return resolvedContext.provider.verifyJwtToken(request.getToken().getToken()).onItemOrFailure()
+                .transformToUni(new BiFunction<TokenVerificationResult, Throwable, Uni<? extends SecurityIdentity>>() {
```
That looks synchronous to me? Do I miss something?

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcProvider.java:
```diff
@@ -0,0 +1,260 @@
+    }
+
+    public Uni<TokenVerificationResult> verifyJwtToken(String token) {
+        JwtConsumerBuilder builder = new JwtConsumerBuilder();
```
That's synchronous, just return the TokenVerificationResult, not wrapped in a Uni

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcProvider.java:
```diff
@@ -0,0 +1,260 @@
+
+    public void introspectToken(UniEmitter<? super TokenVerificationResult> emitter, String token) {
+        try {
+            emitter.complete(introspectToken(token).await().indefinitely());
```
You should use a timeout.

cescoffier replied:
I believe it would be simpler to make `introspectToken` synchronous, and make sure they are always called from a thread that can be blocked.

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcProvider.java:
```diff
@@ -0,0 +1,260 @@
+    public void getUserInfo(UniEmitter<? super JsonObject> emitter, RoutingContext vertxContext,
+            TokenAuthenticationRequest request) {
+        try {
+            emitter.complete(getUserInfo(vertxContext, request).await().indefinitely());
```
I don't believe there is a need to block here. 
Why not returning the Uni directly?

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcProvider.java:
```diff
@@ -0,0 +1,260 @@
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.RoutingContext;
+
+public class OidcProvider {
```
This class contains a lot of `await` wrapped in emitters.

Either we need to say: ok, it's blocking and then remove all the emitters, or we chain our Uni correctly as I don't see anything that can't be made full non-blocking.

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRecorder.java:
```diff
@@ -273,4 +190,105 @@ public void setSecurityEventObserved(boolean isSecurityEventObserved) {
+                    .expireIn(expireInDelay);
+        }
+        return metadataUni.onItemOrFailure()
+                .transformToUni(new BiFunction<OidcConfigurationMetadata, Throwable, Uni<? extends OidcProviderClient>>() {
```
This is synchronous, use `transform` instead.

sberyozkin replied:
Hi @cescoffier, yeah, I've already used `transform` alone a few times, but the reason I've used `transformToUni` here is that the exception at the start up is easier to follow. 
If we have `transform`:
```
Caused by: io.smallrye.mutiny.CompositeException: Multiple exceptions caught:
	[Exception 0] java.lang.IllegalStateException: Retries exhausted : 1 attempts against 1614178544958/1614178542736 expiration
	[Exception 1] io.quarkus.oidc.OIDCException: OIDC server is not available at the 'https://localhost:8543/auth/realms/quarkus' URL. Please make sure it is correct. Note it has to end with a realm value if you work with Keycloak, for example: 'https://localhost:8180/auth/realms/quarkus'
	at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureMap$UniOnItemOrFailureMapProcessor.onFailure(UniOnItemOrFailureMap.java:59)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromPublisher$PublisherSubscriber.onError(UniCreateFromPublisher.java:79)
...
io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.onFailure(UniToMultiPublisher.java:86)
	at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onFailure(UniOperatorProcessor.java:44)
	at io.smallrye.mutiny.vertx.AsyncResultUni.lambda$subscribe$1(AsyncResultUni.java:36)
	at io.vertx.mutiny.ext.web.client.HttpRequest$9.handle(HttpRequest.java:619)
        ....
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: io.quarkus.oidc.OIDCException: OIDC server is not available at the 'https://localhost:8543/auth/realms/quarkus' URL. Please make sure it is correct. Note it has to end with a realm value if you work with Keycloak, for example: 'https://localhost:8180/auth/realms/quarkus'
		at io.quarkus.oidc.runtime.OidcRecorder.toOidcException(OidcRecorder.java:196)
		at io.quarkus.oidc.runtime.OidcRecorder$3.apply(OidcRecorder.java:265)
		at io.quarkus.oidc.runtime.OidcRecorder$3.apply(OidcRecorder.java:260)
		at io.smallrye.context.impl.wrappers.SlowContextualBiFunction.apply(SlowContextualBiFunction.java:21)
		at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureMap$UniOnItemOrFailureMapProcessor.onFailure(UniOnItemOrFailureMap.java:54)
		... 97 more
	Caused by: java.lang.IllegalStateException: Retries exhausted : 1 attempts against 1614178544958/1614178542736 expiration
		at io.smallrye.mutiny.helpers.ExponentialBackoff.lambda$randomExponentialBackoffFunctionExpireAt$2(ExponentialBackoff.java:113)
		at io.smallrye.context.impl.wrappers.SlowContextualFunction.apply(SlowContextualFunction.java:21)
		at io.smallrye.mutiny.groups.MultiOnItem.lambda$transformToUni$7(MultiOnItem.java:354)
		at io.smallrye.mutiny.operators.multi.MultiFlatMapOp$FlatMapMainSubscriber.onItem(MultiFlatMapOp.java:176)
		... 75 more
	Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8543
	Caused by: java.net.ConnectException: Connection refused
		at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
		...
Caused by: [CIRCULAR REFERENCE: java.lang.IllegalStateException: Retries exhausted : 1 attempts against 1614178544958/1614178542736 expiration]
``` 

with `transformToUni`:
```
Caused by: io.quarkus.oidc.OIDCException: OIDC server is not available at the 'https://localhost:8543/auth/realms/quarkus' URL. Please make sure it is correct. Note it has to end with a realm value if you work with Keycloak, for example: 'https://localhost:8180/auth/realms/quarkus'
	at io.quarkus.oidc.runtime.OidcRecorder.toOidcException(OidcRecorder.java:196)
	at io.quarkus.oidc.runtime.OidcRecorder$3.apply(OidcRecorder.java:265)
	at io.quarkus.oidc.runtime.OidcRecorder$3.apply(OidcRecorder.java:260)
	at io.smallrye.context.impl.wrappers.SlowContextualBiFunction.apply(SlowContextualBiFunction.java:21)
	at io.smallrye.mutiny.operators.uni.UniOnItemOrFailureFlatMap$UniOnItemOrFailureFlatMapProcessor.performInnerSubscription(UniOnItemOrFailureFlatMap.java:86)
...
 io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.onFailure(UniToMultiPublisher.java:86)
	at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onFailure(UniOperatorProcessor.java:44)
	at io.smallrye.mutiny.vertx.AsyncResultUni.lambda$subscribe$1(AsyncResultUni.java:36)
	at io.vertx.mutiny.ext.web.client.HttpRequest$9.handle(HttpRequest.java:619)
	...
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Retries exhausted : 1 attempts against 1614178770278/1614178768070 expiration
	at io.smallrye.mutiny.helpers.ExponentialBackoff.lambda$randomExponentialBackoffFunctionExpireAt$2(ExponentialBackoff.java:113)
	at io.smallrye.context.impl.wrappers.SlowContextualFunction.apply(SlowContextualFunction.java:21)
	at io.smallrye.mutiny.groups.MultiOnItem.lambda$transformToUni$7(MultiOnItem.java:354)
	at io.smallrye.mutiny.operators.multi.MultiFlatMapOp$FlatMapMainSubscriber.onItem(MultiFlatMapOp.java:176)
	... 75 more
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8543
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	...
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

There are fewer nested traces in the latter case and no `CIRCULAR_REFERENCE`. This is only reason I've kept it as `transformToUni` for now

sberyozkin replied:
I don't really mind much, I thought it was a bit easier to get the actual error in the latter case...

cescoffier replied:
You mean it's creating a CompositeException with the failure and the thrown exception. 
So, yes, using Uni avoids this.

@jponge, I got that comment twice in two days. In onItemOfFailure, should we really build a composite exception?

jponge replied:
@cescoffier I created a quick issue, we'll investigate this https://github.com/smallrye/smallrye-mutiny/issues/488

cescoffier replied:
thanks @jponge

sberyozkin replied:
@cescoffier @jponge OK, thanks :-)

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:
```diff
@@ -562,67 +562,50 @@ public void accept(Throwable throwable) {
-                        }
-                    });
-                }
+                resolver.getBlockingExecutor().execute(new Runnable() {
```
@cescoffier cool, I was not aware of this simpler pattern, going to update all the similar calls

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:
```diff
@@ -137,29 +136,30 @@ public SecurityIdentity apply(SecurityIdentity identity) {
-                            throw AuthenticationRedirectException.class.cast(throwable);
+                    public Uni<? extends SecurityIdentity> apply(Throwable t) {
+                        if (t instanceof AuthenticationRedirectException) {
+                            throw AuthenticationRedirectException.class.cast(t);
```
will do

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:
```diff
@@ -105,24 +107,30 @@
-                    public Uni<SecurityIdentity> apply(TokenVerificationResult result) {
+                    public Uni<SecurityIdentity> apply(TokenVerificationResult result, Throwable t) {
+                        if (t != null) {
+                            return Uni.createFrom().failure(new AuthenticationFailedException(t));
```
@cescoffier Yeah, I think I've tried it before and retried, the problem with the direct throw is that it causes `500` as opposed to `401` etc, I think `HttpSecurityRecorder` may need to be tweaked a bit to ensure this exception is still recognized, but I'd rather follow up on it later on, otherwise I suspect I'd have hard time rebasing this PR against the Vert.x 4 branch :-)

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:
```diff
@@ -105,24 +107,30 @@
-                    public Uni<SecurityIdentity> apply(JsonObject userInfo) {
+                    public Uni<SecurityIdentity> apply(JsonObject userInfo, Throwable t) {
+                        if (t != null) {
+                            return Uni.createFrom().failure(new AuthenticationFailedException(t));
```
Please see below

sberyozkin commented on extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientRecorder.java:
```diff
@@ -114,36 +115,44 @@ protected static OidcClient createOidcClient(OidcClientConfig oidcConfig, String
         }
-        return tokenRequestUriUni.onItem().transform(new Function<String, OidcClient>() {
+        return tokenRequestUriUni.onItemOrFailure()
+                .transformToUni(new BiFunction<String, Throwable, Uni<? extends OidcClient>>() {
```
Please see below (`OidcRecorder` comment)

sberyozkin replied:
@cescoffier there is no CompositeException in `OidcClientRecorder` so I've changed to `transform` (will commit shortly), will double check again `OidcRecorder`

sberyozkin commented:
Hi @cescoffier thanks for continuing educating me on how to use Uni better :-), I've managed to get rid of most of the emitters, and also simplified `onFailure` handlers. It is getting cleaner for sure. I've resolved the comments which I think I've addressed - please reopen if you think more work is needed.

The only 2 categories of comments which have not been addressed are:
- `return Uni.createFrom().failure()` is still used - probably can be be avoided but some vertx extension security code may need to be tweaked and if possible I'd like to avoid it in this PR
- `transformToUni()` is used for now in `OidcRecorder`/`OidcClientRecorder` to get an arguably shorter exception messages in case of the connection timeout 

let me know please what you think, I may be missing something...

thanks

pedroigor reviewed (approved):
I trust our tests. If we break something we should be able to increase coverage :)

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:
```diff
@@ -562,67 +562,32 @@ public void accept(Throwable throwable) {
-            }
-        });
+        if (BlockingOperationControl.isBlockingAllowed()) {
+            return configContext.provider.refreshTokens(refreshToken);
```
[Optional]

you can write something like:

```
return configContext.provider.refreshTokens(refreshToken)
     .plug(u -> {
         if (! BlockingOperationControl.isBlockingAllowed()) {
             return u.runSubscriptionOn(resolver.getBlockingExecutor());
         } 
         return u;
     });
```

sberyozkin replied:
@cescoffier Looks good to me - it feels more atomic this way, instead of having to duplicate the same call like `refreshTokens` twice, I'll give it a try

sberyozkin replied:
This is done

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/JsonWebKeyCache.java:
```diff
@@ -0,0 +1,48 @@
+    private static final String EC_KEY_TYPE = "EC";
+    private static final String SIGNATURE_USE = "sig";
+
+    private Map<String, Key> keys = new HashMap<>();
```
Aren't you afraid of an out-of-memory issue? 
Nobody removes anything from the map.

sberyozkin replied:
@cescoffier Hi Clement - good question and I was  not sure immediately what was the right answer :-). It is safe - this map never grows.
We get this Map on the start up - it represents a JWK set - which is typically a single key in providers like Keycloak, Auth0, etc, a few max. It never grows - and when Keycloak rotates the keys and Quarkus fetches a new map - this map is replaced as opposed to augmented, please see [here](https://github.com/quarkusio/quarkus/pull/14953/files#diff-c5749413dbc9d20dc1c0e93ed83597f458c909f60eff7ca7b1ad641fe6edc1f9R190).
Thanks for spotting this Map :-), it is worth checking indeed

cescoffier reviewed (changes requested):
Just a quick question about the key cache.

sberyozkin commented:
@cescoffier @gsmet @stuartwdouglas @pedroigor So far JDK15 build has failed with a reactive resteasy test so I'd rather not to restart as the CI is overloaded anyway :-). If Native Security tests pass then I'll be happy to merge it

sberyozkin commented:
Well, I've been waiting for the whole day for it to finish but it did not, so have to restart

sberyozkin commented:
@gsmet @stuartwdouglas @pedroigor @cescoffier Unfortunately the build failed again, Kubernetes Standard test, but Native Security2 is good, the build has been consistently green for several rounds before yesterday, I did a minor cleanup to optimize a Uni sequence involving `.plug` based on Clement's feedback, all is good there, so I'll merge now and will start looking at `vertx4-snapshot`
